src/neo/Consensus/ConsensusService.cs

```diff
@@ -185,8 +187,41 @@ private void InitializeConsensus(byte viewNumber)
             {
                 ChangeTimer(TimeSpan.FromMilliseconds(Blockchain.MillisecondsPerBlock << (viewNumber + 1)));
             }
+
+            // Try to speed up consensus if cached future payloads exists
+            if (context.HasFuturePayloads())
+            {
+                TryToConsumeFuturePayload(context.FuturePreparationPayloads);
+                if (context.HasFuturePayloads())
+                    TryToConsumeFuturePayload(context.FutureCommitPayloads);
+                if (context.HasFuturePayloads())
+                    TryToConsumeFuturePayload(context.FutureChangeViewPayloads);
+                if (context.HasFuturePayloads())
+                    TryToConsumeFuturePayload(context.FutureRecoveryPayloads);
+            }
         }
 
+        private void TryToConsumeFuturePayload(ConsensusPayload[] payloadsArray)
+        {
+            for (int p = 0; p < payloadsArray.Length; p++)
+            {
+                var payload = payloadsArray[p];
+                if (payload != null)
+                {
+                    if (payload.BlockIndex < context.Block.Index)
+                    {
+                        // This future payload is not future anymore, discard it
+                        payloadsArray[p] = null;
+                    }
+                    else if (payload.PrevHash == context.Block.PrevHash && payload.BlockIndex == context.Block.Index)
+                    {
+                        ReverifyAndProcessPayload(payload);
+                        payloadsArray[p] = null;
+                    }
+                    context.CountFuturePayloads = context.CountFuturePayloads - 1;
```

Comment:
If `payload.BlockIndex > context.Block.Index`, we need to decrease the count too?

Reply:
Nooo...aheuahuea 
This is wrong, let's make it inside the `If`. Nice catch.

Reply:
Fixed.